File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/analyticsevents/AnalyticsEventsHandler.java
package uk.nhs.nhsx.analyticsevents;

import uk.nhs.nhsx.core.Environment;
import uk.nhs.nhsx.core.EnvironmentKeys;
import uk.nhs.nhsx.core.HttpResponses;
import uk.nhs.nhsx.core.SystemClock;
import uk.nhs.nhsx.core.UniqueId;
import uk.nhs.nhsx.core.auth.ApiName;
import uk.nhs.nhsx.core.auth.Authenticator;
import uk.nhs.nhsx.core.auth.ResponseSigner;
import uk.nhs.nhsx.core.aws.s3.AwsS3Client;
import uk.nhs.nhsx.core.aws.s3.BucketName;
import uk.nhs.nhsx.core.aws.s3.ObjectKeyNameProvider;
import uk.nhs.nhsx.core.aws.s3.PartitionedObjectKeyNameProvider;
import uk.nhs.nhsx.core.aws.s3.S3Storage;
import uk.nhs.nhsx.core.events.Events;
import uk.nhs.nhsx.core.events.PrintingJsonEvents;
import uk.nhs.nhsx.core.routing.Routing;
import uk.nhs.nhsx.core.routing.RoutingHandler;

import java.time.Instant;
import java.util.function.Supplier;

import static uk.nhs.nhsx.core.StandardSigning.signResponseWithKeyGivenInSsm;
import static uk.nhs.nhsx.core.auth.StandardAuthentication.awsAuthentication;
import static uk.nhs.nhsx.core.routing.Routing.Method.POST;
import static uk.nhs.nhsx.core.routing.Routing.path;
import static uk.nhs.nhsx.core.routing.StandardHandlers.authorisedBy;
import static uk.nhs.nhsx.core.routing.StandardHandlers.withSignedResponses;

public class AnalyticsEventsHandler extends RoutingHandler {

    private final Routing.Handler handler;

    @SuppressWarnings("unused")
    public AnalyticsEventsHandler() {
        this(Environment.fromSystem(), SystemClock.CLOCK, new PrintingJsonEvents(SystemClock.CLOCK));
    }

    public AnalyticsEventsHandler(Environment environment, Supplier<Instant> clock, Events events) {
        this(environment,
            awsAuthentication(ApiName.Mobile, events),
            signResponseWithKeyGivenInSsm(environment, clock, events),
            new AwsS3Client(events),
            new PartitionedObjectKeyNameProvider(clock, UniqueId.ID),
            events,
            awsAuthentication(ApiName.Health, events));
    }

    public AnalyticsEventsHandler(Environment environment,
                                  Authenticator authenticator,
                                  ResponseSigner signer,
                                  S3Storage s3Storage,
                                  ObjectKeyNameProvider objectKeyNameProvider,
                                  Events events,
                                  Authenticator healthAuthenticator) {
        BucketName bucketName = environment.access.required(EnvironmentKeys.SUBMISSION_STORE);
        boolean acceptRequestsEnabled = environment.access.required(Environment.EnvironmentKey.bool("ACCEPT_REQUESTS_ENABLED"));

        this.handler = withSignedResponses(
            events,
            environment,
            signer,
            Routing.routes(
                authorisedBy(authenticator,
                    path(POST, "/submission/mobile-analytics-events", request -> {
                        if (!acceptRequestsEnabled) {
                            return HttpResponses.serviceUnavailable();
                        }

                        var payload = new PayloadValidator().maybeValidPayload(request.getBody());
                        if (payload.isEmpty()) {
                            return HttpResponses.badRequest();
                        }
                        new AnalyticsEventsSubmissionService(s3Storage, objectKeyNameProvider, bucketName, events).accept(payload.get());
                        return HttpResponses.ok();
                    })
                ),
                authorisedBy(healthAuthenticator,
                    path(POST, "/submission/mobile-analytics-events/health", (r) ->
                        HttpResponses.ok()
                    )
                )
            )
        );
    }

    @Override
    public Routing.Handler handler() {
        return handler;
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/analyticsevents/AnalyticsEventsHandler.kt
package uk.nhs.nhsx.analyticsevents

import com.amazonaws.services.kms.AWSKMSClientBuilder
import uk.nhs.nhsx.core.Environment
import uk.nhs.nhsx.core.EnvironmentKeys
import uk.nhs.nhsx.core.HttpResponses.badRequest
import uk.nhs.nhsx.core.HttpResponses.ok
import uk.nhs.nhsx.core.HttpResponses.serviceUnavailable
import uk.nhs.nhsx.core.StandardSigningFactory
import uk.nhs.nhsx.core.SystemClock
import uk.nhs.nhsx.core.UniqueId
import uk.nhs.nhsx.core.auth.ApiName.Health
import uk.nhs.nhsx.core.auth.ApiName.Mobile
import uk.nhs.nhsx.core.auth.Authenticator
import uk.nhs.nhsx.core.auth.ResponseSigner
import uk.nhs.nhsx.core.auth.StandardAuthentication.awsAuthentication
import uk.nhs.nhsx.core.aws.s3.AwsS3Client
import uk.nhs.nhsx.core.aws.s3.ObjectKeyNameProvider
import uk.nhs.nhsx.core.aws.s3.PartitionedObjectKeyNameProvider
import uk.nhs.nhsx.core.aws.s3.S3Storage
import uk.nhs.nhsx.core.aws.ssm.AwsSsmParameters
import uk.nhs.nhsx.core.events.Events
import uk.nhs.nhsx.core.events.PrintingJsonEvents
import uk.nhs.nhsx.core.routing.ApiGatewayHandler
import uk.nhs.nhsx.core.routing.Routing.Method.POST
import uk.nhs.nhsx.core.routing.Routing.path
import uk.nhs.nhsx.core.routing.Routing.routes
import uk.nhs.nhsx.core.routing.RoutingHandler
import uk.nhs.nhsx.core.routing.StandardHandlers.authorisedBy
import uk.nhs.nhsx.core.routing.StandardHandlers.withSignedResponses
import java.time.Instant
import java.util.function.Supplier

class AnalyticsEventsHandler @JvmOverloads constructor(
    environment: Environment = Environment.fromSystem(),
    clock: Supplier<Instant> = SystemClock.CLOCK,
    events: Events = PrintingJsonEvents(clock),
    authenticator: Authenticator = awsAuthentication(Mobile, events),
    signer: ResponseSigner = StandardSigningFactory(
        clock,
        AwsSsmParameters(),
        AWSKMSClientBuilder.defaultClient()
    ).signResponseWithKeyGivenInSsm(environment, events),
    s3Storage: S3Storage = AwsS3Client(events),
    objectKeyNameProvider: ObjectKeyNameProvider = PartitionedObjectKeyNameProvider(clock, UniqueId.ID),
    healthAuthenticator: Authenticator = awsAuthentication(Health, events)
) : RoutingHandler() {

    private val handler = withSignedResponses(
        events,
        environment,
        signer,
        routes(
            authorisedBy(
                authenticator,
                path(POST, "/submission/mobile-analytics-events",
                    ApiGatewayHandler { request, _ ->
                        when {
                            environment.access.required(Environment.EnvironmentKey.bool("ACCEPT_REQUESTS_ENABLED")) -> {
                                val payload = PayloadValidator().maybeValidPayload(request.body)
                                payload?.let {
                                    AnalyticsEventsSubmissionService(
                                        s3Storage,
                                        objectKeyNameProvider,
                                        environment.access.required(EnvironmentKeys.SUBMISSION_STORE),
                                        events
                                    ).accept(payload)
                                    ok()
                                } ?: badRequest()
                            }
                            else -> serviceUnavailable()
                        }
                    })
            ),
            authorisedBy(
                healthAuthenticator,
                path(POST, "/submission/mobile-analytics-events/health", ApiGatewayHandler { _, _ -> ok() })
            )
        )
    )

    override fun handler() = handler
}
